README.md

[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)

# package.rari <img src="./tools/rari_logo.png" alt="rari logo" align="right" height="138.5" style="display: inline-block;">

`rari` is a Python implementation of Pinto et. al's ranked adjusted Rand index (RARI) from
[Ranked Adjusted Rand: integrating distance and partition information in a measure of clustering agreement](https://doi.org/10.1186/1471-2105-8-44).
RARI is an extension of the [adjusted Rand index](https://en.wikipedia.org/wiki/Rand_index) (ARI)
that measures the agreement between two independent clustering solutions while incorporating distances
between instances/clusters from each solution.

* **RARI = 1:** Perfect agreement between cluster solutions 'A' and 'B'. Identical cluster partitions and equally
*ranked* relative distances between clusters in cluster solutions 'A' and 'B'.

* **RARI = 0:** No agreement between cluster solutions 'A' and 'B'. Only occurs when, in cluster solution
'A', all instances are in the same cluster and, in cluster solution 'B', all instances are in their own cluster and all
clusters are equidistant from each other.

Roughly speaking, the benefit of RARI is in penalizing the ARI when a given pair of instances is close together in cluster
solution 'A' and far apart in cluster solution 'B'.

## Lightning Example

* Below is a comparison of the agreement between hierarchical and k-means clustering solutions on the iris data set. The
same distance matrix is used to calculate pairwise distances between each iris instance, but this is not a requirement.

``` python
from sklearn.datasets import load_iris
from sklearn.cluster import AgglomerativeClustering, KMeans
from sklearn.metrics import pairwise_distances
from rari import rari

X = load_iris().data

model_1 = AgglomerativeClustering(n_clusters=3, linkage='ward')
x = model_1.fit_predict(X)

model_2 = KMeans(n_clusters=3)
y = model_2.fit_predict(X)

dist_x = pairwise_distances(X, metric='euclidean')
dist_y = pairwise_distances(X, metric='euclidean')

rari(x, y, dist_x, dist_y)
```
Out[1]: **.975**

## Install

* Development

``` python
pip install git+https://github.com/nredell/rari
```

## Intuition

Below is Figure 1 from Pinto et. al's article which demonstrates the impact of inter-cluster distances on the RARI
metric as compared to, say, the ARI.

<p align="center">
  <img src="tools/figure_1.PNG" width="400px"></img>
</p>

## Examples

### Example 1: ARI vs. RARI, Few Clusters, High Agreement

``` python
import numpy as np
import pandas as pd
from sklearn.datasets import make_blobs
from sklearn.cluster import AgglomerativeClustering, KMeans
from sklearn.metrics import adjusted_rand_score, pairwise_distances
from rari import rari

X, y = make_blobs(n_samples=[50, 50, 50], n_features=2, cluster_std=1.0, center_box=(-5.0, 5.0), shuffle=True, random_state=224)
data = pd.DataFrame(np.hstack([X, y[:, np.newaxis]]), columns=["X1", "X2", "Cluster"])

model_1 = AgglomerativeClustering(n_clusters=3, linkage='ward')
x = model_1.fit_predict(X)

model_2 = KMeans(n_clusters=3)
y = model_2.fit_predict(X)

dist_x = pairwise_distances(X, metric='euclidean')
dist_y = pairwise_distances(X, metric='euclidean')
```

![](./tools/example_1_plot_1.png) ![](./tools/example_1_plot_2.png) ![](./tools/example_1_plot_3.png)

``` python
adjusted_rand_score(x, y)
rari(x, y, dist_x, dist_y)
```
**ARI:** .83
**RARI:** .89

### Example 2: ARI vs. RARI, A New Data Point

The toy 1D example below illustrates how the dynamic RARI changes as the distance between clusters changes
while the static ARI remains the same.

Imagine that the moving data point represents a new data point added
to the data set, at which point each of 2 models is re-run and the clusters are re-labeled. For the sake of illustration, the
labels for this new data point from each model are held constant through each of the 11 analyses to emphasize the impact of cluster spacing.
In a real problem, it's likely that the moving data point would be classified as a '2' as it approaches the yellow '2' on the
right hand side of each plot. However, this change of labels may not even occur in a simple 2D example with a method like spectral
clustering. And our intuitions will fail us in higher dimensions, but RARI will account for these changes in cluster orientation if so desired.

![](./tools/rari_points.gif)

## Implementation Details

At present, inter-cluster distances are based on the euclidean distance between pairs of instances in `dist_x` and `dist_y`.
That is to say, even if the input pairwise distance matrices are, for example, cosine and manhattan, the inter-cluster distance ranks
are still based on a euclidean, complete linkage measure of these pairwise distances. This will be relaxed in the future with support
for additional input arguments.



HISTORY.rst
=======
History
=======

0.1.0 (2020-03-18)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for rari."""


tests/test_rari.py
#!/usr/bin/env python

"""Tests for `rari` package."""


import unittest
from click.testing import CliRunner

from rari import rari
from rari import cli


class TestRari(unittest.TestCase):
    """Tests for `rari` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_000_something(self):
        """Test something."""

    def test_command_line_interface(self):
        """Test the CLI."""
        runner = CliRunner()
        result = runner.invoke(cli.main)
        assert result.exit_code == 0
        assert 'rari.cli.main' in result.output
        help_result = runner.invoke(cli.main, ['--help'])
        assert help_result.exit_code == 0
        assert '--help  Show this message and exit.' in help_result.output


rari/rari.py
"""Main module."""

import numpy as np
from scipy.stats import rankdata
from pandas import Categorical, crosstab

def rari(x, y, dist_x, dist_y):

    """ Computes the ranked adjusted Rand index for measuring clustering agreement.

        An implementation of Pinto et. al's ranked adjusted Rand index which is an extension of the adjusted Rand index
        that measures the agreement between two independent clustering solutions while incorporating distances
        between instances/clusters in each solution. An index of 1 indicates perfect agreement between solutions while
        and index close to 0 indicates random labeling and distances.

        Pinto, F. R., Carriço, J. A., Ramirez, M., & Almeida, J. S. (2007). Ranked Adjusted Rand: integrating distance
            and partition information in a measure of clustering agreement. BMC Bioinformatics, 8(1), 44.
            https://doi.org/10.1186/1471-2105-8-44

        Parameters
        ----------
        x : 1D np.array
            A numeric vector of cluster membership values with 1 value per instance. Cluster values should start at either
            0 or 1. The number of clusters can differ between 'x' and 'y'; however, len(x) must equal len(y).

        y : 1D np.array
            A numeric vector of cluster membership values with 1 value per instance. Cluster values should start at either
            0 or 1. The number of clusters can differ between 'x' and 'y'; however, len(x) must equal len(y).

        dist_x : 2D np.array
            A len(x) by len(x) matrix giving the distances between instances in 'x'. The distance metric used to
            create 'dist_x' can be different from that used to create 'dist_y'.

        dist_y : 2D np.array
            A len(y) by len(y) matrix giving the distances between instances in 'y'. The distance metric used to
            create 'dist_y' can be different from that used to create 'dist_x'.

        Returns
        ----------
        rari : float
        """
    #-------------------------------------------------------------------------
    assert all([isinstance(x, np.ndarray), isinstance(y, np.ndarray), isinstance(dist_x, np.ndarray),
                isinstance(dist_y, np.ndarray)]), \
                "'x', 'y', 'dist_x', and 'dist_y' must all be of type 'numpy.ndarray'."
    assert x.shape == y.shape, "'x' and 'y' should be 1D numpy arrays--e.g., (100, )--with the same length."
    assert dist_x.shape == dist_y.shape, "'dist_x' and 'dist_y' should be square matrices--e.g., (100, 100)--with the same dimensions."
    assert dist_x.shape[0] == dist_x.shape[1] and dist_y.shape[0] == dist_y.shape[1], "'dist_x' and 'dist_y' should be square matrices--e.g., (100, 100)--with the same dimensions."

    n_instances = x.size

    clusters_x = np.sort(np.unique(x))
    clusters_y = np.sort(np.unique(y))

    # Cluster membership should be a numeric vector starting at 0 or 1.
    if 0 not in clusters_x:
        clusters_x = clusters_x - 1

    if 0 not in clusters_y:
        clusters_y = clusters_y - 1

    n_clusters_x = len(clusters_x)
    n_clusters_y = len(clusters_y)
    # -------------------------------------------------------------------------------
    # Determine which instances in x and y belong to which cluster.
    clust_instances_x = []
    for i in clusters_x:
        clust_instances_x.append(np.arange(n_instances)[i == x])

    clust_instances_y = []
    for i in clusters_y:
        clust_instances_y.append(np.arange(n_instances)[i == y])
    # -------------------------------------------------------------------------------
    # Separately for each cluster solution 'x' and 'y', for each instance in the input
    # distance matrix, compute the average distance between the select instance and
    # all instances in each cluster. The result is an n_instance by n_cluster matrix.
    # TODO: Add additional methods--beyond mean()--for computing pairwise distances.
    clust_dist_x = []
    for cluster_instances in clust_instances_x:
        for instance in range(n_instances):
            clust_dist_x.append(np.nanmean(dist_x[instance, cluster_instances]))
    clust_dist_x = np.array(clust_dist_x).reshape(n_instances, -1, order='F')

    clust_dist_y = []
    for cluster_instances in clust_instances_y:
        for instance in range(n_instances):
            clust_dist_y.append(np.nanmean(dist_y[instance, cluster_instances]))
    clust_dist_y = np.array(clust_dist_y).reshape(n_instances, -1, order='F')

    clust_dist_x = np.hstack([clust_dist_x, x[:, np.newaxis]])
    clust_dist_y = np.hstack([clust_dist_y, y[:, np.newaxis]])
    # -------------------------------------------------------------------------------
    # Eqn 4., Table 3. rmm is the rank mismatch matrix indexed from 0 to number of clusters -1.

    # The algorithm proceeds in two steps:

    # First, compute the n_clusters by n_clusters matrix of distances between clusters for each
    # cluster solution x and y. The distance matrix may be asymmetric. As a result, the matrix
    # should be interpreted as "the cluster in column j is [this distance] from the cluster in
    # row i."
    clust_dist_matrix_x = []
    for column in range(n_clusters_x):
        for cluster_instances in clust_instances_x:
            clust_dist_matrix_x.append(np.nanmean(clust_dist_x[cluster_instances, column]))
    clust_dist_matrix_x = np.array(clust_dist_matrix_x).reshape(n_clusters_x, n_clusters_x, order='C')

    clust_dist_matrix_y = []
    for column in range(n_clusters_y):
        for cluster_instances in clust_instances_y:
            clust_dist_matrix_y.append(np.nanmean(clust_dist_y[cluster_instances, column]))
    clust_dist_matrix_y = np.array(clust_dist_matrix_y).reshape(n_clusters_y, n_clusters_y, order='C')

    # As the inter-cluster distances will be rank ordered, all instances in a cluster have a 0 inter-cluster distance.
    np.fill_diagonal(clust_dist_matrix_x, 0)
    np.fill_diagonal(clust_dist_matrix_y, 0)

    # Rank the distances between clusters. This matrix may be asymmetric. Subtracting 1 for easier indexing.
    for i in range(n_clusters_x):
        clust_dist_matrix_x[:, i] = rankdata(clust_dist_matrix_x[:, i], method='dense') - 1

    for i in range(n_clusters_y):
        clust_dist_matrix_y[:, i] = rankdata(clust_dist_matrix_y[:, i], method='dense') - 1

    # Second, compute instance-level pairwise differences in cluster distance ranks and
    # set the resulting n_instance by n_instance diagonal to NA to remove the bias in calculating
    # an instances distance from itself.
    instance_dist_rank_matrix_x = []
    for i in range(n_instances):
        for j in range(n_instances):
            if i != j:  # Heaviside function.
                instance_dist_rank_matrix_x.append(clust_dist_matrix_x[x[i], x[j]])
            else:
                instance_dist_rank_matrix_x.append(np.nan)

    instance_dist_rank_matrix_y = []
    for i in range(n_instances):
        for j in range(n_instances):
            if i != j:  # Heaviside function.
                instance_dist_rank_matrix_y.append(clust_dist_matrix_y[y[i], y[j]])
            else:
                instance_dist_rank_matrix_y.append(np.nan)

    rmm = crosstab(Categorical(instance_dist_rank_matrix_x), Categorical(instance_dist_rank_matrix_y)).values
    # -------------------------------------------------------------------------------
    # Eqn 5. mdd is the mean diagonal deviance. Note: the Rand index = 1 - mdd.
    mdd_numerator = []
    for i in range(rmm.shape[0]):
        for j in range(rmm.shape[1]):
            mdd_numerator.append(rmm[i, j] * abs(i / rmm.shape[0] - j / rmm.shape[1]))
    mdd = sum(mdd_numerator) / (n_instances ** 2 - n_instances)
    # -------------------------------------------------------------------------------
    # Eqn 6. rmmi is the expected value of the rank mismatch matrix under independence.
    rmmi = np.zeros((rmm.shape[0], rmm.shape[1]))
    for i in range(rmm.shape[0]):
        for j in range(rmm.shape[1]):
            rmmi[i, j] = np.sum(rmm[i, :]) * (np.sum(rmm[:, j]) / np.sum(rmm))
    # -------------------------------------------------------------------------------
    # Eqn 5. mddi is the mean diagonal deviance under independence.
    mddi_numerator = []
    for i in range(rmm.shape[0]):
        for j in range(rmm.shape[1]):
            mddi_numerator.append(rmmi[i, j] * abs(i / rmmi.shape[0] - j / rmmi.shape[1]))
    mddi = sum(mddi_numerator) / (n_instances ** 2 - n_instances)
    # -------------------------------------------------------------------------------
    # Eqn. 7.
    rari = (mddi - mdd) / mddi

    return rari
